extensions/grpc/stubs/src/main/java/io/quarkus/grpc/stubs/MultiStreamObserver.java

```diff
@@ -1,28 +1,62 @@
 package io.quarkus.grpc.stubs;
 
-import io.grpc.stub.StreamObserver;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import io.grpc.stub.ClientCallStreamObserver;
+import io.grpc.stub.ClientResponseObserver;
 import io.smallrye.mutiny.subscription.MultiEmitter;
 
-public class MultiStreamObserver<T> implements StreamObserver<T> {
+public class MultiStreamObserver<I, O> implements ClientResponseObserver<I, O> {
+    private final MultiEmitter<? super O> emitter;
+
+    // If this is set, the Multi was terminated by completion or failure, not by cancellation.
+    private final AtomicBoolean terminated = new AtomicBoolean();
+
+    private AtomicReference<ClientCallStreamObserver<I>> requestStreamObserver = new AtomicReference<>();
```

Comment:
You might be able to condense the 2 atomic variables into one (`requestStreamObserver`, when `null` it means terminated)

Reply:
Is it guaranteed that `beforeStart` is called in every case? If not, the emitter will never be terminated on error (after the change, `emitter.fail()` is only called if `requestStreamObserver` is not `null`).